### PR TITLE
event listening improvements

### PR DESCRIPTION
**Pull request by JustynaBroniszewska**

### CODE REVIEW

File: universal-login-example/src/services/IdentityService.js [hunk 1]
```diff
@@ -25,20 +25,19 @@ class IdentityService {
       this.identity.address
     );
     const {address} = new Wallet(this.privateKey);
+    const filter = {contractAddress: this.identity.address, key: address};
     this.subscription = this.sdk.subscribe(
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/IdentityService.js]: How about a oneliner?
`this.subscription = this.sdk.subscribe('KeyAdded', filter, () => {`

File: universal-login-example/src/services/IdentityService.js [hunk 2]
```diff
@@ -25,20 +25,19 @@ class IdentityService {
...
+      filter,
+      () => {
+        this.cancelSubscription();
+        this.identity = {
+          name: this.identity.name,
+          privateKey: this.privateKey,
+          address: this.identity.address
+        };
+        this.emitter.emit('setView', 'Greeting', {greetMode: 'addKey'});
+        this.storeIdentity(this.identity);
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/IdentityService.js]: extract method?

File: universal-login-example/src/services/IdentityService.js [hunk 3]
```diff
@@ -47,20 +46,19 @@ class IdentityService {
...
+      filter,
+      () => {
+        this.cancelSubscription();
+        this.identity = {
+          name: this.identity.name,
+          privateKey: this.privateKey,
+          address: this.identity.address
+        };
+        this.emitter.emit('setView', 'Greeting');
+        this.storeIdentity(this.identity);
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/IdentityService.js]: extract method

File: universal-login-sdk/lib/observers/ObserverBase.js [hunk 4]
```diff
@@ -8,9 +8,15 @@ class ObserverBase {
...
-    this.emitters[identityAddress] = emitter;
+  subscribe(eventType, filter, callback) {
+    let emitter;      
+    emitter = this.emitters[filter.contractAddress] || new EventEmitter();
+    this.emitters[filter.contractAddress] = emitter;
+    if (filter.key) {
+      this.emitters[filter.contractAddress] ? this.emitters[filter.contractAddress] : this.emitters[filter.contractAddress] = {};
+      emitter = this.emitters[filter.contractAddress][filter.key] || new EventEmitter();
+      this.emitters[filter.contractAddress][filter.key] = emitter;
+    }
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/observers/ObserverBase.js]: Let's use here linear search

File: universal-login-sdk/lib/observers/BlockchainObserver.js [hunk 5]
```diff
@@ -16,22 +16,27 @@ class BlockchainObserver extends ObserverBase {
...
-    const filter = {fromBlock: this.lastBlock, address: identityAddress, topics};
+    const filter = {fromBlock: this.lastBlock, address: contractAddress, topics};
     const events = await this.provider.getLogs(filter);
     for (const event of events) {
-      this.emitters[identityAddress].emit(type, this.parseArgs(type, event));
+      const {address} = this.parseArgs(type, event);
+      if (this.emitters[contractAddress][address]) {
+        this.emitters[contractAddress][address].emit(type, this.parseArgs(type, event));
+      } else {
+        this.emitters[contractAddress].emit(type, this.parseArgs(type, event));
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/observers/BlockchainObserver.js]: Let's use here linear search

File: universal-login-example/src/services/IdentityService.js (hunk 1 above)
Comment (JustynaBroniszewska) [reply universal-login-example/src/services/IdentityService.js]: ✅

File: universal-login-example/src/services/IdentityService.js (hunk 2 above)
Comment (JustynaBroniszewska) [reply universal-login-example/src/services/IdentityService.js]: ✅

File: universal-login-example/src/services/IdentityService.js (hunk 3 above)
Comment (JustynaBroniszewska) [reply universal-login-example/src/services/IdentityService.js]: ✅

File: universal-login-sdk/lib/observers/ObserverBase.js (hunk 4 above)
Comment (JustynaBroniszewska) [reply universal-login-sdk/lib/observers/ObserverBase.js]: ✅

File: universal-login-sdk/lib/observers/BlockchainObserver.js (hunk 5 above)
Comment (JustynaBroniszewska) [reply universal-login-sdk/lib/observers/BlockchainObserver.js]: ✅

File: universal-login-example/src/services/IdentityService.js [hunk 6]
```diff
@@ -21,48 +21,33 @@ class IdentityService {
...
-        }
-      }
-    );
+    const filter = {contractAddress: this.identity.address, key: address};
+    this.subscription = this.sdk.subscribe('KeyAdded', filter, () => {
+      this.eventService();
+    });
+  }
+
+  eventService(options = {}) {
```
Comment (marekkirejczyk) [inline universal-login-example/src/services/IdentityService.js]: viewOptions

**Pull request merged**